### PR TITLE
Dbw node: Drive by wire component working (fixes #7)

### LINKED ISSUE #7: Build an initial version of the Controller
Move the car along the waypoints at a constant speed disregarding the speed information contained in the input waypoints.

### PR DESCRIPTION
**Pull request by buffos**
An initial version of the drive by wire node. 
It works fine. Things that can be tuned

* Number of waypoints to use to create a polynomial path representation (current = 10)
* Degree of polynomial to use (current = 3)
* Pid parameters for both steering_pid and throttle_pid
* Configure Filters. Now a 90-10 interpolation of New-Old values is used

### CODE REVIEW

**Comment by buffos** [comment]
So how it does not want the matplotlib import and some unused variables
( removed their usage for now, but they might be used later)

How do I fix it now. comment those and do a new PR?

**Comment by buffos** [comment]
Also what to do with matplotlib?

**Comment by mithi** [comment]
@buffos 

I think this is the one causing the error
```
E: 11, 0: Unable to import 'matplotlib.pyplot' (import-error)
```
that was only used for testing right? So maybe you can remove it :)

**Comment by mithi** [comment]
When you push to your branch it updates the pull request. That's what happened last time I made a pull request :)

**Comment by buffos** [comment]
Grrrrrr. I told it to ignore import-error. I think I placed it in the wrong position

**Comment by buffos** [comment]
!!! I do not understand the message. Unable to import dbw_helper? Why?

**Comment by mithi** [comment]
@buffos 
Just a trailing white space now! :)
```
C:102, 0: Trailing whitespace (trailing-whitespace)
```

**Comment by buffos** [comment]
LOL
The problem is I pylint does not run on the VM, so I have to do workarounds

**Comment by buffos** [comment]
Finally 🥇

File: ros/src/twist_controller/dbw_helper.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+# Date:   28.08.17
+#-------------------------------------------------------------------------------
+"""
+A collection of helper functions for dbw node
+"""
+# pylint: disable=import-error
+from math import sqrt, cos, sin
+import numpy as np
+import matplotlib.pyplot as plt
+import tf
```
Comment (asimonov) [inline ros/src/twist_controller/dbw_helper.py:13]: name 'tf' may be misleading when we start using tensorflow

Comment (buffos) [reply ros/src/twist_controller/dbw_helper.py:13]: its the name of the package. its not mine :)

Comment (buffos) [reply ros/src/twist_controller/dbw_helper.py:13]: besides you will not be using it in that same file

File: ros/src/twist_controller/dbw_helper.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    Args:
+        pose (object) : A pose object
+        waypoints (list) : A list of waypoint objects
+
+    Returns:
+        cte (float) : the cross track error (signed)
+    """
+    x_coords, y_coords = shift_and_rotate_waypoints(pose, waypoints, POINTS_TO_FIT)
+    coefficients = np.polyfit(x_coords, y_coords, 3)
+    distance = np.polyval(coefficients, 2.0)
```
Comment (asimonov) [inline ros/src/twist_controller/dbw_helper.py:197]: should we not use fit_polynomial defined above?

Comment (buffos) [reply ros/src/twist_controller/dbw_helper.py:197]: we use it.. We get the coefficients and use them.

Comment (buffos) [reply ros/src/twist_controller/dbw_helper.py:197]: Oh no.. That is another function that may be deleted. I see what you mean

Comment (buffos) [reply ros/src/twist_controller/dbw_helper.py:197]: In this function I want to select only few waypoints...

Comment (buffos) [reply ros/src/twist_controller/dbw_helper.py:197]: The other function just gets waypoints. Here I transform the coordinate system, select few and then fit

File: ros/src/twist_controller/dbw_node.py [hunk 3]
```diff
@@ -27,72 +27,129 @@
...
+                         self.twist_cb, queue_size=1)
 
         self.loop()
 
     def loop(self):
-        #pylint: disable=no-self-use
         """
-        STUB
+        The main functionallity of the DBW Node.
+        It runs at 50Hz = 50 cycles per second, until ROS shuts down.
```
Comment (asimonov) [inline ros/src/twist_controller/dbw_node.py:119]: runs at 10Hz after most recent change

Comment (buffos) [reply ros/src/twist_controller/dbw_node.py:119]: I guess this is not fixed. I just changed that for now, for my VM. 
Will probably change later

**Review by mithi — DISMISSED** [review]
Changes in `waypoint_updater`:
- helper functions placed in separate file

Changes in `twist_controller`:
- helper functions placed in separate file
- Implement preliminary controller:
--> get current velocity, current pose, target velocity and position, waypoints
---> get cte given current pose and waypoints
---> get velocity_error from current and target velocity
---> use pid and lowpass filter to calculate throttle/brake and steering angle based in cte and velocity error

File: ros/src/twist_controller/dbw_node.py (hunk 3 above)
Comment (buffos) [reply ros/src/twist_controller/dbw_node.py:119]: If you run it on a native installation, you can change it back to 50 (i guess) and see if it works

**Comment by buffos** [comment]
What is the correct action now?
Merge? Rebase and merge? or Squash and merge?
And does this action refer to this repo, or my forked one?

File: ros/src/twist_controller/twist_controller.py [hunk 4]
```diff
@@ -1,35 +1,93 @@
...
+
+    def control(self, vel_error, cte, dbw_enabled):
         """
-        STUB
+        Given then target linear and angular velocities, as well as the current velocity
+        calculate the CTE and pass them to the corresponding PID's
+
+        Args:
+             vel_error (float) : Velocity Error
+             current_velocity (float) : The current linear velocity in the x axis
```
Comment (asimonov) [inline ros/src/twist_controller/twist_controller.py]: I think you removed that parameter

**Review by asimonov — COMMENTED** [review]
just a few comments

File: ros/src/twist_controller/twist_controller.py (hunk 4 above)
Comment (buffos) [reply ros/src/twist_controller/twist_controller.py]: oops! correct.

Comment (buffos) [reply ros/src/twist_controller/twist_controller.py]: fixed

Comment (buffos) [reply ros/src/twist_controller/twist_controller.py]: I really like this interface!!!

File: ros/src/twist_controller/dbw_helper.py (hunk 1 above)
Comment (asimonov) [reply ros/src/twist_controller/dbw_helper.py:13]: ok, lets leave it :)

File: ros/src/twist_controller/dbw_helper.py (hunk 2 above)
Comment (asimonov) [reply ros/src/twist_controller/dbw_helper.py:197]: ok. just wanted to double check we are not duplicating code

**Review by asimonov — APPROVED** [review]
looks good!

File: ros/src/twist_controller/dbw_helper.py (hunk 1 above)
Comment (ljanyst) [reply ros/src/twist_controller/dbw_helper.py:13]: Well, that's how they named the package. Not much can be done about it. Also, I believe they were first ;)

File: ros/src/twist_controller/dbw_helper.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+# Author: Kostas Oreopoulos <kostas.oreopoulos@gmail.com>
+# Date:   28.08.17
+#-------------------------------------------------------------------------------
+"""
+A collection of helper functions for dbw node
+"""
+# pylint: disable=import-error
+from math import sqrt, cos, sin
+import numpy as np
+import matplotlib.pyplot as plt
```
Comment (ljanyst) [inline ros/src/twist_controller/dbw_helper.py:12]: There is no place for this here. This requires xserver and won't run in headless mode when configured this way. Also, we're only supposed to use the dependencies that are in the requirements file and nothing else.

Common practice to do this sort of debugging is to dump data to a file and have a separate plotting script.

**Review by ljanyst — CHANGES_REQUESTED** [review]

**Comment by ljanyst** [comment]
I was unable to successfully run the code. This is what, I get:

```
Traceback (most recent call last):
  File "/home/ljanyst/Apps/python/envs/ros/local/lib/python2.7/site-packages/engineio/server.py", line 398, in _trigger_event
    return self.handlers[event](*args)
  File "/home/ljanyst/Apps/python/envs/ros/local/lib/python2.7/site-packages/socketio/server.py", line 513, in _handle_eio_message
    self._handle_event(sid, pkt.namespace, pkt.id, pkt.data)
  File "/home/ljanyst/Apps/python/envs/ros/local/lib/python2.7/site-packages/socketio/server.py", line 449, in _handle_event
    self._handle_event_internal(self, sid, data, namespace, id)
  File "/home/ljanyst/Apps/python/envs/ros/local/lib/python2.7/site-packages/socketio/server.py", line 452, in _handle_event_internal
    r = server._trigger_event(data[0], namespace, sid, *data[1:])
  File "/home/ljanyst/Apps/python/envs/ros/local/lib/python2.7/site-packages/socketio/server.py", line 481, in _trigger_event
    return self.handlers[namespace][event](*args)
  File "/home/ljanyst/Projects/carla-driver/ros/src/styx/server.py", line 32, in telemetry
    bridge.publish_odometry(data)
  File "/home/ljanyst/Projects/carla-driver/ros/src/styx/bridge.py", line 136, in publish_odometry
    pose = self.create_pose(data['x'], data['y'], data['z'], data['yaw'])
KeyError: 'z'
message handler error
Traceback (most recent call last):
  File "/home/ljanyst/Apps/python/envs/ros/local/lib/python2.7/site-packages/engineio/server.py", line 398, in _trigger_event
    return self.handlers[event](*args)
  File "/home/ljanyst/Apps/python/envs/ros/local/lib/python2.7/site-packages/socketio/server.py", line 513, in _handle_eio_message
    self._handle_event(sid, pkt.namespace, pkt.id, pkt.data)
  File "/home/ljanyst/Apps/python/envs/ros/local/lib/python2.7/site-packages/socketio/server.py", line 449, in _handle_event
    self._handle_event_internal(self, sid, data, namespace, id)
  File "/home/ljanyst/Apps/python/envs/ros/local/lib/python2.7/site-packages/socketio/server.py", line 452, in _handle_event_internal
    r = server._trigger_event(data[0], namespace, sid, *data[1:])
  File "/home/ljanyst/Apps/python/envs/ros/local/lib/python2.7/site-packages/socketio/server.py", line 481, in _trigger_event
    return self.handlers[namespace][event](*args)
  File "/home/ljanyst/Projects/carla-driver/ros/src/styx/server.py", line 32, in telemetry
    bridge.publish_odometry(data)
  File "/home/ljanyst/Projects/carla-driver/ros/src/styx/bridge.py", line 136, in publish_odometry
    pose = self.create_pose(data['x'], data['y'], data['z'], data['yaw'])
TypeError: 'NoneType' object has no attribute '__getitem__'
```

**Comment by ljanyst** [comment]
OK, please disregard my previous comment. The problem was caused by a wrong version of the simulator. It works fine:

[Test #01](https://www.youtube.com/watch?v=jzTqgxONuQY)

**Comment by ljanyst** [comment]
After a couple of circles, I get this:
```
Traceback (most recent call last):
  File "/home/ljanyst/Projects/carla-driver/ros/src/twist_controller/dbw_node.py", line 195, in <module>
    DBWNode()
  File "/home/ljanyst/Projects/carla-driver/ros/src/twist_controller/dbw_node.py", line 114, in __init__
    self.loop()
  File "/home/ljanyst/Projects/carla-driver/ros/src/twist_controller/dbw_node.py", line 131, in loop
    cte = dbw_helper.cte(self.pose, self.waypoints)
  File "/home/ljanyst/Projects/carla-driver/ros/src/twist_controller/dbw_helper.py", line 194, in cte
    x_coords, y_coords = shift_and_rotate_waypoints(pose, waypoints, POINTS_TO_FIT)
  File "/home/ljanyst/Projects/carla-driver/ros/src/twist_controller/dbw_helper.py", line 142, in shift_and_rotate_waypoints
    shift_x = waypoints[i].pose.pose.position.x - originX
IndexError: list index out of range
```

**Review by ljanyst — APPROVED** [review]

**Pull request merged**